src/main/java/jenkins/plugins/slack/SlackUserIdResolver.java

```diff
@@ -0,0 +1,156 @@
+            RunWithSCM r = (RunWithSCM) run;
+            return resolveUserIdsForChangeLogSets(r.getChangeSets());
+        } else {
+            return Collections.EMPTY_LIST;
+        }
+    }
+
+    public List<String> resolveUserIdsForChangeLogSet(ChangeLogSet changeLogSet) {
+        return Arrays.stream(changeLogSet.getItems())
+                .map(item -> resolveUserId(((Entry) item).getAuthor()))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    public List<String> resolveUserIdsForChangeLogSets(List<ChangeLogSet> changeLogSets) {
+        return changeLogSets.stream()
+                .map(changeLogSet -> resolveUserIdsForChangeLogSet(changeLogSet))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    public List<String> resolveUserIdsForEmailAddresses(List<String> emailAddresses) {
+        return emailAddresses.stream()
+                .map(address -> resolveUserIdForEmailAddress(address))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    public String resolveUserId(User user) {
+        for (MailAddressResolver resolver : mailAddressResolvers) {
+            String emailAddress = resolver.findMailAddressFor(user);
+            if (StringUtils.isNotEmpty(emailAddress)) {
+                String userId = resolveUserIdForEmailAddress(emailAddress);
+                if (StringUtils.isNotEmpty(userId)) {
+                    return userId;
+                }
+            }
+        }
+        return null;
+    }
+
+    public String resolveUserIdForEmailAddress(String emailAddress) {
+        if (StringUtils.isEmpty(emailAddress) || StringUtils.isEmpty(authToken)) {
+            return null;
+        }
+
+        // prepare get method for looking up Slack userId by email address
+        String url = String.format(LOOKUP_BY_EMAIL_METHOD_URL_FORMAT, authToken, emailAddress);
+
+        HttpGet getRequest = new HttpGet(url);
+        try {
+            CloseableHttpResponse response = httpClient.execute(getRequest);
+            int responseCode = response.getStatusLine().getStatusCode();
+            // inspect the response content if a 200 response code is received
+            if (HttpStatus.SC_OK == responseCode) {
+                HttpEntity entity = response.getEntity();
+                JSONObject slackResponse = new JSONObject(EntityUtils.toString(entity));
+                // additionally, make sure the JSON response contains an 'ok: true' entry
+                if (slackResponse.optBoolean(SLACK_OK_FIELD)) {
+                    JSONObject slackUser = slackResponse.getJSONObject(SLACK_USER_FIELD);
+                    return slackUser.getString(SLACK_ID_FIELD);
+                }
+            }
+        } catch (IOException | ParseException | JSONException ex) {
+            LOGGER.log(Level.WARNING, "Error getting userId from Slack", ex);
```

Comment:
It would be good if this went in the build log as well, because of how restricted jenkins logs usually are this sort of error doesn't normally get surfaced, I would suggest throwing the exception up and catching it one level higher and logging it

Reply:
I agree with you about visibility of the warning logs. However, I don't quite follow your suggestion - as I see it, I would have to throw the exception up to the slackSend pipeline step / slack publisher level in order to have access to a `TaskListener` for logging.

Reply:
@timja I believe this is the only unresolved conversation left. Perhaps this can be addressed in a follow-up PR?

Reply:
I don't see any code that handles rate limiting :sweat_smile:

Reply:
I think it's probably worth either adding a `SlackNotificationsLogger` or a PrintStream to the build log, shouldn't need the whole taskListener